Add metrics for database service

Tener opened the pull request:
See:
- https://github.com/gravitational/teleport/issues/26667.

Changelog: Add several metrics for Proxy and Database services.

Tener commented:
All, kindly suggest other metrics you'd like to see implemented.

Docs PR will follow. I'd like to finalize the list of metrics before opening one.

strideynet commented on lib/srv/db/common/report.go:
```diff
@@ -0,0 +1,246 @@
+var (
+	clientReadBytes = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "db_client_read_bytes",
```
Generally, Prometheus recommends: 

> that an accumulating count has total as a suffix, in addition to the unit if applicable.

Source: https://prometheus.io/docs/practices/naming/

Tener replied:
Done.

rosstimothy reviewed (commented):
Please add all of the new metrics to the teleport namespace.

smallinsky commented on lib/srv/db/common/engines.go:
```diff
@@ -130,3 +148,18 @@ func (c *EngineConfig) CheckAndSetDefaults() error {
+func (c *EngineConfig) ObserveConnectionSetupTime() func() {
+	start := c.Clock.Now()
+	return func() {
+		connectionSetupTime.WithLabelValues(teleport.ComponentDatabase, c.Name).Observe(c.Clock.Since(start).Seconds())
```
solely database name label won't  give us to much information especially for cloud customers. Could we add the database type and distinguish the real type like RDS slefhosted, AWS hosted  ?

Tener replied:
`name` is the name of the engine, not the database. Do you think we should add the actual database name here too?

smallinsky commented on lib/srv/db/common/engines.go:
```diff
@@ -130,3 +148,18 @@ func (c *EngineConfig) CheckAndSetDefaults() error {
+	}
+}
+
+func (c *EngineConfig) IncMessagesFromClient() {
+	messagesFromClient.WithLabelValues(teleport.ComponentDatabase, c.Name).Inc()
+}
+
+func (c *EngineConfig) IncMessagesFromServer() {
+	messagesFromServer.WithLabelValues(teleport.ComponentDatabase, c.Name).Inc()
+}
```
Does these function really belongs to EngineConfig ?

smallinsky commented on lib/srv/db/common/auth.go:
```diff
@@ -1023,3 +1023,28 @@ func (r *elastiCacheRedisIAMTokenRequest) getSignableRequest() (*http.Request, e
+	return r.Auth.GetTLSConfig(ctx, sessionCtx)
+}
+
+func (r *reportingAuth) methodCallMetrics() func() {
+	return methodCallMetrics(r.component, r.databaseProtocol, 2)
+}
```
The `methodCallMetrics` is too generic and referring to runtime runtime stacktrace many times seems to be bad ideal. We we cannot just forward a func name ?

Tener replied:
We are picking up the function name from the stacktrace, and by using the stacktrace info we don't have to manually provide the function names everywhere. Are you worried about the performance impact here? I'm pretty sure this is rather fast and on par with `trace.Wrap` which is everywhere.

Tener replied:
I've dropped all but 2 calls, which makes it feasible to pass the method name explicitly without runtime information, which removes the code in question.

smallinsky commented on lib/srv/db/common/auth.go:
```diff
@@ -1023,3 +1023,28 @@ func (r *elastiCacheRedisIAMTokenRequest) getSignableRequest() (*http.Request, e
+
+func (r *reportingAuth) GetTLSConfig(ctx context.Context, sessionCtx *Session) (*tls.Config, error) {
+	defer r.methodCallMetrics()()
+	return r.Auth.GetTLSConfig(ctx, sessionCtx)
```
What if this call return err ? The invocation time metrics will be misleading.

Tener replied:
What do you propose should we do in `err` case?

smallinsky commented on lib/srv/db/common/audit.go:
```diff
@@ -80,12 +90,18 @@ func NewAudit(config AuditConfig) (Audit, error) {
+
 // OnSessionStart emits an audit event when database session starts.
 func (a *audit) OnSessionStart(ctx context.Context, session *Session, sessionErr error) {
+	defer a.methodCallMetrics()()
```
the `methodCallMetrics` is also called in bellow in `a.EmitEvent(ctx, event)`
What do we want to measure here ?

Tener replied:
Yeah, I wondered about it too. I landed on the current design because both methods are callable independently and the actual metrics incremented are different because their method names are different. They will be highly correlated, though.

Tener replied:
We now only measure `EmitEvent`

smallinsky commented on lib/srv/db/common/report.go:
```diff
@@ -0,0 +1,246 @@
+		update := func() {
+			tx, rx := trackingClientConn.Stat()
+			clientWrittenBytes.WithLabelValues(r.component, r.engineName).Add(float64(tx - txPrev))
+			clientReadBytes.WithLabelValues(r.component, r.engineName).Add(float64(rx - rxPrev))
```
The `clientReadBytes`  metrics actually aggregate all reads operation from all connections per engine. 

I wonder  what we are gaining having this low level network metric per clients pool ?

Tener replied:
We can make it more granular if we'd like. Would you like to see any particular labels added?

On its own, without any additional information, it probably isn't very useful, but on the other hand, many other metrics are like this. It can be very powerful when answering certain questions, though. 

For example: given one client, one Teleport instance, and one database server, what is the maximum throughput that can be achieved? How does this answer change if we increase the latency anywhere? How do different protocols compare? Does compression helps or not?

smallinsky reviewed (commented):
First pass. 
Could we also add metrics to measure connection time from teleport proxy to db agent ?

Tener commented:
@smallinsky 
> Could we also add metrics to measure connection time from teleport proxy to db agent ?

Added.

Tener commented:
All, I've address the PR feedback I got and rebased on master. Passing DB type required some refactoring, but also showed me the fragility of `.WithLabelValues()` method, which is why there are helper methods now to make future changes easier.

evanfreed commented on lib/srv/db/common/reporting.go:
```diff
@@ -0,0 +1,272 @@
+	connectionSetupTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: teleport.MetricNamespace,
+			Name:      "db_connection_setup_time",
```
```suggestion
			Name:      "db_connection_setup_time_seconds",
```

nit: suffix describing unit ([reference](https://prometheus.io/docs/practices/naming/#metric-names)).

evanfreed commented on lib/srv/db/common/reporting.go:
```diff
@@ -0,0 +1,272 @@
+	activeConnections = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: teleport.MetricNamespace,
+			Name:      "db_active_connections",
```
```suggestion
			Name:      "db_active_connections_total",
```

nit: suffix describing unit ([reference](https://prometheus.io/docs/practices/naming/#metric-names)).

evanfreed reviewed (commented):
Thanks for the additions, much appreciated 👍 I looked at prometheus naming style, only nits.

smallinsky commented on lib/srv/db/common/reporting.go:
```diff
@@ -0,0 +1,272 @@
+	return &reportingEngine{cfg}, nil
+}
+
+func (r *reportingEngine) InitializeConnection(clientConn net.Conn, sessionCtx *Session) error {
```
I would suggest to drop this metrics for now or add this on the teleport proxy side instead of teleport db agent side to have more insight in Teleport Cloud.

Tener replied:
Ok, dropped.

smallinsky commented on lib/srv/db/common/reporting.go:
```diff
@@ -0,0 +1,272 @@
+		commonLabels,
+	)
+
+	clientWrittenBytes = prometheus.NewCounterVec(
```
Could you also add these metrics to our metrics docs: https://github.com/gravitational/teleport/blob/871aa2af1b953e4d18e528b1bef2ae324efa2e6d/docs/pages/includes/metrics.mdx#L107

Tener replied:
Yes, I'll raise a separate docs-only PR once this one lands (and metric names are settled etc.).

Tener commented on lib/srv/db/common/reporting.go:
```diff
@@ -0,0 +1,272 @@
+	_ = metrics.RegisterPrometheusCollectors(prometheusCollectorsEngine...)
+}
+
+func NewReportingEngine(cfg reporterConfig) (Engine, error) {
```
Fixed.

Tener commented on lib/srv/db/common/reporting.go:
```diff
@@ -0,0 +1,272 @@
+	closer := utils.NewCloserConn(clientConn)
+	trackingClientConn := utils.NewTrackingConn(closer)
+
+	go func() {
```
No longer relevant since these metrics were dropped per Marek's suggestion.

Tener commented on lib/srv/db/proxyserver.go:
```diff
@@ -622,3 +645,88 @@ func getConfigForClient(conf *tls.Config, ap auth.ReadDatabaseAccessPoint, log l
+			Buckets:   prometheus.LinearBuckets(1, 1, 16),
+		},
+		commonLabels,
+	)
```
Done.

smallinsky reviewed (approved):
the PR LGTM but I think that we are still missing:
- metrics docs update
- conn count metrics from proxy side.

Tener commented:
> the PR LGTM but I think that we are still missing:
> 
> * metrics docs update
> * conn count metrics from proxy side.

Thanks, these two items will go into separate PRs.

@evanfreed @rosstimothy PTAL.

Tener commented:
@Joerger do you mind re-approving this PR?

Tener committed:
Add DB metrics

Tener commented on lib/srv/db/common/auth.go:
```diff
@@ -1033,3 +1033,24 @@ func (r *elastiCacheRedisIAMTokenRequest) getSignableRequest() (*http.Request, e
+	db        types.Database
+}
+
+func GetReportingAuth(db types.Database, auth Auth) Auth {
```
Do you mean the advice to return a concrete type, which means `reportingAuth` here?

Tener replied:
Fixed.

Tener commented on lib/srv/db/common/auth.go:
```diff
@@ -1033,3 +1033,24 @@ func (r *elastiCacheRedisIAMTokenRequest) getSignableRequest() (*http.Request, e
+	return r.Auth.GetTLSConfig(ctx, sessionCtx)
+}
+
+var _ Auth = (*reportingAuth)(nil)
```
This expresses the intent: `*reportingAuth` should implement `Auth`. This is trivially true now, but it may not be in the future if someone touches that code.

Tener replied:
Yes, this is true now. Yet a simple change may invalidate it:

```patch
type reportingAuth struct {
-	          Auth
+	auth      Auth
	component string
	db        types.Database
}
```

This is just a test; only it is written using the type system.

Tener commented on lib/srv/db/common/reporting.go:
```diff
@@ -0,0 +1,243 @@
+	return trace.Wrap(r.engine.HandleConnection(ctx, sessionCtx))
+}
+
+var _ Engine = &reportingEngine{}
```
This is a static assertion. We have a number of those already and they work nicely to express the intent.

Tener commented on lib/srv/db/common/reporting.go:
```diff
@@ -0,0 +1,243 @@
+	}
+}
+
+func (r *reportingEngine) getLabels() prometheus.Labels {
+	return getLabels(r.component, r.database)
+}
```
The savings are minuscule, but yeah, we can cache it.

Tener commented on lib/srv/db/common/reporting.go:
```diff
@@ -0,0 +1,243 @@
+)
+
+func methodCallMetrics(method, component string, db types.Database) func() {
+	start := time.Now()
```
We are not using the clock in truth. I'll drop it in favor of plain `time`.

Tener committed:
Address PR feedback.

- rename functions to better indicate their intent
- drop unused clock field
- cache labels
- add godoc

Tener committed:
Merge branch 'master' into tener/db-metrics

Tener committed:
Merge branch 'master' into tener/db-metrics

Tener committed:
Refactoring to return concrete types instead of interfaces.

Tener commented on lib/srv/db/common/reporting.go:
```diff
@@ -0,0 +1,243 @@
+}
+
+// NewReportingEngine returns new reporting engine, which wraps a regular Engine but reports various usage metrics.
+func NewReportingEngine(cfg reporterConfig) (Engine, error) {
```
Fixed.
